### PR TITLE
Additional places where locations can be avoided

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Small bits of cleanup in tehse simple cases:

image

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 1]
```diff
@@ -2052,14 +2052,14 @@ private BoundExpression BindNonMethod(SimpleNameSyntax node, Symbol symbol, Bind
...
-                {
-                    return node.SpanStart < localSymbolLocation.SourceSpan.Start;
-                }
+                // trivial position check, before more costly tree check (which requires walking up the nodes). Most
+                // code is correct, so this check is expected to succeed nearly every time.
+                if (node.SpanStart >= declarator.SpanStart)
+                    return false;
 
-                return false;
+                return node.SyntaxTree == declarator.SyntaxTree;
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:2062]: Here, I am wondering how we would possibly think this `node` might be a use of `localSymbol`, unless it is already in the same syntax tree. No need to attempt to change logic as part of eliminating Location usages, though.

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs [hunk 2]
```diff
@@ -1170,7 +1170,7 @@ protected virtual void ReportUnassigned(Symbol symbol, SyntaxNode node, int slot
 
             if (skipIfUseBeforeDeclaration &&
                 symbol.Kind == SymbolKind.Local &&
-                (symbol.Locations.Length == 0 || node.Span.End < symbol.GetFirstLocationOrNone().SourceSpan.Start))
+                (symbol.TryGetFirstLocation() is var location && (location is null || node.Span.End < location.SourceSpan.Start)))
```
Comment (sharwell) [inline src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs:1173]: 📝 This is going to conflict with #67881 (where `TryGetFirstLocation` is eliminated)

Comment (sharwell) [inline src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs:1173]: I'm not a fan of treating a canonical `Try*` failure path as a sneaky true in this expression. Does this work?

```suggestion
                (symbol.TryGetFirstLocation() is not { } location || (node.Span.End < location.SourceSpan.Start)))
```

**Comment by CyrusNajmabadi** [comment]
@tmat ptal

**Pull request merged**